providers/common/include/prov/hybrid_pkey.h

```diff
@@ -0,0 +1,84 @@
+/*
+ * Copyright 2024 The OpenSSL Project Authors. All Rights Reserved.
+ *
+ * Licensed under the Apache License 2.0 (the "License").  You may not use
+ * this file except in compliance with the License.  You can obtain a copy
+ * in the file LICENSE in the source distribution or at
+ * https://www.openssl.org/source/license.html
+ */
+
+#include <openssl/evp.h>
+#include "internal/refcount.h"
+
+#define MAX_HYBRID_ALGS 2
+#define MAX_HYBRID_PARAMS 20
+
+typedef struct {
+    const unsigned int num_algs;
+    const struct {
+        const char *name;
+        size_t key_length_bytes;
+        size_t bits;
+        size_t security_bits;
+        size_t shared_secret_bytes;
+        size_t ciphertext_bytes;
+    } alg[MAX_HYBRID_ALGS];
+} HYBRID_ALG_INFO;
+
+#define HYBRID_COMMON_FIELDS            \
+    OSSL_LIB_CTX *libctx;               \
+    char *propq;                        \
```

Comment:
Could you please clarify why do we need this field and why do we have it one but not one per key?

Reply:
fwiw and imo it makes sense to only have one as the "amalgamation" result representing the "joint" algorithm.

Reply:
Corresponding to what? The key algos may come from different providers and have different propq...

Reply:
Corresponding to the constituent algorithms' parameters -- that must be identical irrespective of which provider delivers the implementation.

Reply:
We need to fetch all of the subalgs so we need a propq to use while doing so.  This is also passed through to the subalgs in case they fetch anything.  For example `ossl_hybrid_kmgmt_set_params` does this:

```c
        case PIDX_PKEY_PARAM_PROPERTIES:
            if (!OSSL_PARAM_get_utf8_string(p, &str, SIZE_MAX))
                return 0;
            OPENSSL_free(key->propq);
            key->propq = str;
            /* Fall through */

        default:
            if (!key_subalg_set_param(key->info->num_algs, key->keys, prms))
                return 0;
            break;
```

I hadn't considered a different query for the subalgs.  It is an interesting idea.  Can you think of a use case?  Would we want a different propq for each subalg or just a single one for all subalgs (that is different to the hybrid's propq).  A _nice_ implementation might be tricky but I'm sure we'd be able to brainstorm something.

Reply:
How are these propq's communicated into the keymgmt logic? Doesn't that happen via "OSSL_PKEY_PARAM_PROPERTIES"? Are you suggesting that this logic should (be changed to) be able to parse a (single param) string into two (separate per-algorithm propqs)? If so, which characters are safe to use as separators for the constituent propq's? Also seems to demand some more documentation.

Reply:
> Regardless, not for this PR IMO.

I agree.

BTW, how do you use propq?

Reply:
One complication: where should the hybrid algorithm be located?  If it's in the FIPS provider, it then isn't possible to load anything from the P11 provider (because the FIPS provider loads only from it's internal library context that's unavailable from the outside).  This implies that it would have to be from the default provider…

The propq is used to fetch the sub-algorithms (`EVP_PKEY_CTX_new_from_name` and `EVP_PKEY_CTX_new_from_pkey`).  It is also passed through to the sub-algorithms in case they have to fetch anything.

Reply:
I'm afraid it should be a property of a sub-algorithm then. But we can discuss it separately